data/rule/rule.go

```diff
@@ -0,0 +1,117 @@
+// Package rule contains rule-based heuristic implementations.
+// It is used in the generated code in content.go for disambiguation of languages
+// with colliding extensions based on regexps from Linguist data.
+package rule
+
+import (
+	"regexp"
+)
+
+// Heuristic consist of a number of rules where each, if matches,
+// identifes content as belonging to a programming language(s).
+type Heuristic interface {
+	Matcher
+	Languages() []string
+}
+
+// Matcher checks if the data matches (number of) pattern.
+// Every rule below implements this interface: a rule is matcher that identifies
+// given programming language(s) in case of the match.
+type Matcher interface {
+	Match(data []byte) bool
+}
+
+// languages struct incapsulate data common to every Matcher: all languages
+// that it identifies.
+type languages struct {
+	langs []string
+}
+
+// Languages returns all languages, identified by this Matcher.
+func (l *languages) Languages() []string {
+	return l.langs
+}
+
+// MatchingLanguages is a helper to create new languages.
+func MatchingLanguages(langs ...string) *languages {
+	return &languages{langs}
+}
+
+// Implements a Heuristic.
+type or struct {
```

Comment:
Consider perhaps `orMatcher`, `andMatcher` etc. for these unexported  types? Or make the constructors self-documenting by having them return `Matcher` instead of their concrete type?

    func Or(l Languages, r *regexp.Regexp) Matcher { …

Reply:
I think you could eliminate the dependency on the `regexp` package entirely from this package: A `*regexp.Regexp` satisfies `rule.Matcher` anyway, at which point the `pattern` field could be a `Matcher` and this argument could as well.

Reply:
Addressed in ec54891 - now they return a proper `Heuristic`